lib/astutils.cpp

```diff
@@ -3647,3 +3647,75 @@ bool isUnevaluated(const Token *tok)
 {
     return Token::Match(tok, "alignof|_Alignof|_alignof|__alignof|__alignof__|decltype|offsetof|sizeof|typeid|typeof|__typeof__ (");
 }
+
+static std::set<MathLib::bigint> getSwitchValues(const Token *startbrace, bool &hasDefault)
+{
+    std::set<MathLib::bigint> values;
+    const Token *endbrace = startbrace->link();
+    if (!endbrace)
+        return values;
+
+    hasDefault = false;
+    for (const Token *tok = startbrace->next(); tok && tok != endbrace; tok = tok->next()) {
+        if (Token::simpleMatch(tok, "{") && tok->scope()->type == Scope::ScopeType::eSwitch) {
+            tok = tok->link();
+            continue;
+        }
+        if (Token::simpleMatch(tok, "default")) {
+            hasDefault = true;
+            break;
+        }
+        if (Token::simpleMatch(tok, "case")) {
+            const Token *valueTok = tok->astOperand1();
+            if (valueTok->hasKnownIntValue())
+                values.insert(valueTok->getKnownIntValue());
+            continue;
+        }
+    }
+
+    return values;
+}
+
+bool isExhaustiveSwitch(const Token *startbrace)
+{
+    if (!startbrace || !Token::simpleMatch(startbrace->previous(), ") {") || startbrace->scope()->type != Scope::ScopeType::eSwitch)
+        return false;
+    const Token *rpar = startbrace->previous();
+    if (!rpar)
+        return false;
+    const Token *lpar = rpar->link();
+    if (!lpar)
+        return false;
+
+    const Token *condition = lpar->astOperand2();
+    if (!condition->valueType())
+        return false;
+
+    bool hasDefault{};
+    std::set<MathLib::bigint> switchValues = getSwitchValues(startbrace, hasDefault);
+
+    if (hasDefault)
+        return true;
+
+    if (condition->valueType()->type == ValueType::Type::BOOL) {
+        const auto trueCase = switchValues.find(1);
+        const auto falseCase = switchValues.find(0);
+        return trueCase != switchValues.end() && falseCase != switchValues.end();
+    }
+
+    if (condition->valueType()->isEnum()) {
+        const Scope *typeScope = condition->valueType()->typeScope;
+        const std::vector<Enumerator> &enumList = typeScope->enumeratorList;
+        std::set<MathLib::bigint> declValues;
+        for (const auto &e : enumList) {
+            if (e.value_known)
```

Comment:
I feel this code can be simplified. It's not necessary with declValues.

Reply:
Perhaps just counting how many values in `switchValues` appear in `enumList` and comparing the result to `enumList.size()`?